src/OpenTelemetry.AutoInstrumentation/Loading/AssemblyLoadDetector.cs

```diff
@@ -0,0 +1,67 @@
+// <copyright file="AssemblyLoadDetector.cs" company="OpenTelemetry Authors">
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// </copyright>
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+
+namespace OpenTelemetry.AutoInstrumentation.Loading
+{
+    internal abstract class AssemblyLoadDetector
+    {
+        private HashSet<string> _requiredAssemblies;
+        private int _loadedCount;
+
+        public AssemblyLoadDetector(IEnumerable<string> requiredAssemblies)
+        {
+            _requiredAssemblies = new HashSet<string>(requiredAssemblies);
+
+            AppDomain.CurrentDomain.AssemblyLoad += CurrentDomain_AssemblyLoad;
```

Comment:
Forked the code to test out the approach. I commented out this event handler wiring and tried the code with console and asp.net core app it could collect traces. 

I tired it on smoke test application, it could collect http client trace too. Does this event wiring needed?

Reply:
Without `AspNetCoreInstrumentation` instance these traces are invalid. With this approach it's very important to create instrumentation instance ASAP after assembly load. This instance overrides ASP.NET spans and fills with correct data (tags).